### PR TITLE
Feature istio adapter maglev

### PR DESCRIPTION
**Pull request by trainyao**
### Issues associated with this PR

#1117  

### Solutions
described in issue

### UT result
on the way

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by wangfakang** [comment]
Here is a `maglev` LB analysis [document](https://docs.google.com/document/d/12lgyCW-GmlErr_ihvAO7tMmRe87i70bv2xqe4h2LUz4/edit?usp=sharing) .

**Comment by trainyao** [comment]
shell merge [this PR](https://github.com/mosn/api/pull/9) first, and commit 6009738f294c6fa3203a7448c881d70204d0440b can be reverted.

**Comment by trainyao** [comment]
vendor dir modification in [this PR](https://github.com/mosn/api/pull/9) is also edited, please check

File: pkg/upstream/cluster/loadbalancer.go [hunk 1]
```diff
@@ -333,3 +344,205 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+		chosen = lb.chooseHostFromSegmentTree(index)
+	}
+
+	log.Proxy.Debugf(context.DownstreamContext(), "[lb][maglev] get index %d host %s %s",
+		index, chosen.Hostname(), chosen.AddressString())
+
+	return chosen
+}
+
+func (lb *maglevLoadBalancer) generateChooseHostHash(context types.LoadBalancerContext, hp *v2.HashPolicy) uint64 {
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:452]: could optimized，forexample：
```
type HashPolicy interface {
GetHash(ctx context, key string) (string, error)
}
```

File: pkg/upstream/cluster/loadbalancer.go [hunk 2]
```diff
@@ -333,3 +344,205 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+		}
+
+		if leaf.IsRoot() {
+			break
+		}
+
+		leaf = lb.fallbackSegTree.FindParent(leaf)
+	}
+
+	return host
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:524]: May be not assigned to host.

File: pkg/upstream/cluster/loadbalancer.go [hunk 3]
```diff
@@ -333,3 +344,205 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+		headerValue, err := variable.GetProtocolResource(context.DownstreamContext(), api.HEADER, headerKey)
+
+		if err == nil {
+			hashString := fmt.Sprintf("%s:%s", headerKey, headerValue)
+			hash := getHashByString(hashString)
+			return hash
+		}
+	}
+	if hp.SourceIP != nil {
+		return getHashByAddr(context.DownstreamConnection().RemoteAddr())
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:464]: could use `mosnctx.Get(ctx, types.ContextOriRemoteAddr)`.

File: pkg/module/segmenttree/segmenttree.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
+package segmenttree
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/module/segmenttree/segmenttree.go:18]: license added

File: pkg/config/v2/json_test.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ package v2
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/config/v2/json_test.go]: gofmtted

File: pkg/stream/http/var_test.go [hunk 6]
```diff
@@ -21,6 +21,8 @@ import (
 	"bufio"
 	"bytes"
 	"context"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/stream/http/var_test.go:24]: goimportsed

File: pkg/stream/http2/var_test.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package http2
+
+import (
+	"context"
+	"fmt"
+	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/stream/http2/var_test.go:25]: goimports-ed

File: pkg/variable/protocolres_test.go [hunk 8]
```diff
@@ -32,8 +32,10 @@ const (
 
 func TestGetProtocolResource(t *testing.T) {
 	m := make(map[string]string)
-	m["http_request_path"] = "/http"
-	m["dubbo_request_path"] = "/dubbo"
+	httpKey := string(HTTP1) + "_" + types.VarProtocolRequestPath
+	dobboKey := string(Dubbo) + "_" + types.VarProtocolRequestPath
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/variable/protocolres_test.go]: fixed

File: pkg/variable/protocolres.go [hunk 9]
```diff
@@ -42,20 +43,25 @@ func RegisterProtocolResource(protocol types.ProtocolName, resource api.Protocol
...
-func GetProtocolResource(ctx context.Context, name api.ProtocolResourceName) (string, error) {
+func GetProtocolResource(ctx context.Context, name api.ProtocolResourceName, data ...interface{}) (string, error) {
 	p, ok := mosnctx.Get(ctx, types.ContextKeyDownStreamProtocol).(types.ProtocolName)
 	if !ok {
 		return "", errors.New("get ContextKeyDownStreamProtocol failed.")
 	}
 
 	if v, ok := protocolVar[convert(p, name)]; ok {
-		return GetVariableValue(ctx, v)
+		dataInput := []interface{}{}
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/variable/protocolres.go]: fixed

File: pkg/types/variable.go [hunk 10]
```diff
@@ -48,15 +48,18 @@ const (
...
+	VarProtocolRequestArgPrefix = "request_arg_"
+	VarProtocolRequestArg       = "request_arg"
+	VarProtocolRequestUri       = "request_uri"
+)
+
 // [Protocol]: http
 const (
 	VarHttpRequestMethod = "http_request_method"
 	VarHttpRequestLength = "http_request_length"
-	VarHttpRequestUri    = "http_request_uri"
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/types/variable.go:55]: const added, and marked deprecated

File: pkg/upstream/cluster/loadbalancer.go (hunk 1 above)
Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: fixed

File: pkg/upstream/cluster/loadbalancer.go [hunk 11]
```diff
@@ -333,3 +344,205 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
 	}
 	return true
 }
+
+func newMaglevLoadBalancer(info types.ClusterInfo, set types.HostSet) types.LoadBalancer {
+	var table *maglev.Table
+	var tree *segmenttree.Tree
+	names := []string{}
+	for _, host := range set.Hosts() {
+		names = append(names, host.Hostname())
```
[the remark this answers is not in the document]
Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: fixed

File: pkg/upstream/cluster/loadbalancer.go (hunk 3 above)
Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: fixed

File: pkg/upstream/cluster/loadbalancer.go [hunk 12]
```diff
@@ -333,3 +338,184 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+	if lb.maglev == nil {
+		return nil
+	}
+
+	route, ok := mosnctx.Get(context.DownstreamContext(), types.ContextKeyDownStreamRouter).(api.Route)
+	if !ok {
+		return nil
+	}
+
+	hash := route.RouteRule().Policy().HashPolicy().GenerateHash(context.DownstreamContext())
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:461]: the value of `Policy().HashPolicy()` maybe is nil.

Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: fixed

**Comment by wangfakang** [comment]
1. maglev ring size needs to be set dynamically depends on the number of host.
2. instead of segment tree with RR after hash failure.
3. Simplify the Maglev algorithm (avoid unuse memory or copy).
4. default use ip_hash policy when hash policy is nil.

File: pkg/upstream/cluster/loadbalancer.go [hunk 13]
```diff
@@ -333,3 +336,105 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+	return lb.HostNum(metadata) > 0
+}
+
+func (lb *maglevLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
+	return len(lb.hosts.Hosts())
+}
+
+// chooseHostFromHostList traverse host list to find a healthy host
+func (lb *maglevLoadBalancer) chooseHostFromHostList(index int) types.Host {
+	for i, host := range lb.hosts.Hosts() {
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:430]: This may cause traffic hot on the peer in front of `hosts`.

Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: fixed

File: pkg/upstream/cluster/loadbalancer.go [hunk 14]
```diff
@@ -333,3 +336,118 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+		counterIndex--
+
+		if lb.hosts.Hosts()[counterIndex].Health() {
+			return lb.hosts.Hosts()[counterIndex]
+		}
+	}
+
+	// go right
+	counterIndex = index
+	for counterIndex < hostCount-1 {
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:444]:
```
for counterIndex = index + 1; counterIndex < hostCount; counterIndex++ {
xxxxx
}
```

Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go:444]: seems the same, is it for code beautify?

Comment (wangfakang) [reply pkg/upstream/cluster/loadbalancer.go:444]: > seems the same, is it for code beautify?

yeah.

Comment (trainyao) [reply pkg/upstream/cluster/loadbalancer.go]: done

File: pkg/stream/http/var.go [hunk 15]
```diff
@@ -19,33 +19,43 @@ package http
...
+	pathName      = fmt.Sprintf("%s_%s", protocol.HTTP1, types.VarProtocolRequestPath)
+	uriName       = fmt.Sprintf("%s_%s", protocol.HTTP1, types.VarProtocolRequestUri)
+	argName       = fmt.Sprintf("%s_%s", protocol.HTTP1, types.VarProtocolRequestArg)
+	argPrefixName = fmt.Sprintf("%s_%s", protocol.HTTP1, types.VarProtocolRequestArgPrefix)
+	argIndex      = len(argPrefixName)
+
 	builtinVariables = []variable.Variable{
 		variable.NewBasicVariable(types.VarHttpRequestMethod, nil, requestMethodGetter, nil, 0),
 		variable.NewBasicVariable(types.VarHttpRequestLength, nil, requestLengthGetter, nil, 0),
-		variable.NewBasicVariable(types.VarHttpRequestUri, nil, requestUriGetter, nil, 0),
```
Comment (nejisama) [inline pkg/stream/http/var.go:40]: 这些配置不能改成小写的，否则其他地方要获取用不了了

File: pkg/proxy/downstream.go [hunk 16]
```diff
@@ -333,6 +333,7 @@ func (s *downStream) OnDestroyStream() {}
 // types.StreamReceiveListener
 func (s *downStream) OnReceive(ctx context.Context, headers types.HeaderMap, data types.IoBuffer, trailers types.HeaderMap) {
 	s.downstreamReqHeaders = headers
+	s.context = mosnctx.WithValue(s.context, types.ContextKeyDownStreamHeaders, headers)
```
Comment (nejisama) [inline pkg/proxy/downstream.go:336]: LB里获取headers和route 不用通过ctx，这里应该是没有对齐。LB有一个自己单独的LBCtx，那里可以直接获取

File: pkg/upstream/cluster/loadbalancer.go [hunk 17]
```diff
@@ -333,3 +336,115 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
...
+	}
+	if nameCount == 0 {
+		return mgv
+	}
+
+	maglevM := maglev.SmallM
+	// according to test, 30000 host with testing 1e8 times, hash distribution begins to go wrong,
+	// max=4855, mean=3333.3333333333335, peak-to-mean=1.4565
+	// so use BigM when host >= 30000
+	limit := 30000
```
Comment (nejisama) [inline pkg/upstream/cluster/loadbalancer.go:369]: 这种应该是定义成常量

**Pull request merged**